### ISSUE TITLE
[BUG] Selecting MIDI Keyboard Not Working

### ISSUE THREAD

**Issue by 00J1NX00** [tofodroid/mimi-mod#119]
**Describe the bug**
My MIDI Keyboard (Yamaha PSR E363) shows up in the MIDI Device Config, but clicking the save button does nothing. The keyboard shows up in the selected device section, but the light doesn't change, and no input from the keyboard is detected by the mod. Tried restarting the game, updating keyboard drivers, unplugging the keyboard, and disabling all mods.

Followed the GitHub page's wiki, so if you want to recreate this bug, do that. ¯\_(ツ)_/¯

**Screenshots**
![image_2024-03-30_022208265](https://github.com/tofodroid/mimi-mod/assets/47050773/04265653-80c6-4bbf-8055-2ba552031f55)

**Environment:**
 - OS: [e.g. Windows 10]
 - Mod Version: [4.0.0 (tried it on 4.0.3, but ran into the no sound bug when using my computer keyboard)]
 - Minecraft Version: [1.20.1]
 - Forge Version: [47.2.0]

**Comment by tofodroid** [comment]
Hello!

Thank you so much for the bug report!

I've been focusing on trying to resolve the no sound bug on 4.0.2/4.0.3 (and I just uploaded version 4.0.4 which will hopefully help with that) but should have time to look into this bug later today or tomorrow!

Thanks again for the report!

**Comment by apocris** [comment]
Hello, I just wanted to say that I am currently having this same issue. My situation is near-identical to OP's where my keyboard is detected, but I can't press save to send it to "selected device". The only difference in mine is that I do not see anything underneath the selected device bar like in the picture displayed here.
I realize this is a work in progress, but I just wanted to double up on this issue. It looks like a great mod otherwise!

**Comment by tofodroid** [comment]
I appreciate the additional confirmation of this issue! I am starting to look into this right now!

**Comment by tofodroid** [comment]
Update: I have found the cause of this issue and should hopefully have a new build out in a few hours that will resolve it!

**Comment by tofodroid** [comment]
I have just uploaded version 4.0.5 to [CurseForge](https://www.curseforge.com/minecraft/mc-mods/mimi-mod) and [Modrinth](https://modrinth.com/mod/mimi/versions) that should hopefully resolve this issue!

The CurseForge uploads are still processing (can take up to 24hours 😞) but the Modrinth versions are available now!

Let me know if you still have any issues!

**Comment by 00J1NX00** [comment]
Hi again, thank you for the quick replies. Been busy so I didn't get around to responding.

Unless I'm doing something wrong, the MIDI keyboard still doesn't send input through the in-game instruments. in 4.0.5 the keyboard properly showed up but didn't input notes, and simply clicking the notes worked yet the sound bug was still there.

In 4.0.6 the no sound bug was properly fixed, yet the MIDI keyboard stays the same, with the green light, properly displayed in the grey box, yet now not only does it not input notes, but clicking the X next to the green light will cause the game to freeze. Don't know if that was an issue in 4.0.5 or not, but it is in 4.0.6.

Sorry for the constant issues, it's always possible I'm doing something wrong. Any recommendations or things I should try?

**Comment by tofodroid** [comment]
@00J1NX00 Thanks for the reply and no worries, I can definitely relate to being busy! 😅 Trying to take full advantage of this small moment where I actually have time to work on the mod a bit haha.

One quick thing to check - Do you have the `Sys MIDI Device` toggle enabled and all of the channels enabled on the instrument(s) that you're trying to play MIDI notes on?

![image](https://github.com/tofodroid/mimi-mod/assets/81447370/0897dffc-a9af-47cd-b928-a539b08a1de7)

If you already have that all enabled, or if enabling that doesn't help, would you mind posting your `latest.log` file? That may help me see if I can figure out what's going on.

Thanks!

**Comment by 00J1NX00** [comment]
Sincerely sorry for my mistake. I had those enabled and nothing happened, but I guess disabling and re-enabling the mod fixed it? Just to make a liar out of me, of course lmao.

All's good now, sorry about that! Very epic mod :)

**Comment by 00J1NX00** [comment]
I retract that statement; it did work, but it doesn't work after a game restart, only after a fresh install/enable of the mod after previously being absent from the game. _Then_ it doesn't detect input and freezes when configuring the MIDI device.

Here's the latest log: https://pastebin.com/57kQXieA

**Comment by tofodroid** [comment]
@00J1NX00 Haha no worries at all and thank you for the latest log file! That's definitely a funky issue with it working right after installing the mod but not working after that unless the mod is re-installed. I'll dig deeper into your log file a bit later today when I have some time and see if I can figure out what's causing that. Thanks for your help looking into this issue!

**Comment by tofodroid** [comment]
@00J1NX00 Hello again! I've just uploaded an experimental build to Modrinth that may help with your issues. Feel free to give it a shot and let me know how it works!

https://modrinth.com/mod/mimi/version/1.20.1-4.0.8

If it seems to fix things for you then I'll package it up into a full release!

**Comment by tofodroid** [comment]
Hello again! I've gone ahead and packaged that fix into a full release. Let me know if you continue to run into any issues!

**Comment by tofodroid** [comment]
Hello again! I'm closing this as resolved for now, but if you continue to experience any problems feel free to comment or open a new issue!
